File: apps/designer/src/components/SrujaCanvas/textMeasure.ts
/**
 * Text Measurement Utility
 *
 * Canvas-based text measurement for accurate node sizing.
 * Uses OffscreenCanvas when available for better performance.
 */

// Cache for text measurements to avoid repeated calculations
const measureCache = new Map<string, { width: number; height: number }>();

// Font configuration interface
interface FontConfig {
  family: string;
  size: number;
  weight: number;
  lineHeight: number;
}

// Font configurations matching node styles
export const FONT_CONFIG: Record<"title" | "technology" | "description", FontConfig> = {
  title: {
    family: "Inter, -apple-system, BlinkMacSystemFont, sans-serif",
    size: 14,
    weight: 600,
    lineHeight: 1.3,
  },
  technology: {
    family: "Inter, -apple-system, BlinkMacSystemFont, sans-serif",
    size: 11,
    weight: 500,
    lineHeight: 1.2,
  },
  description: {
    family: "Inter, -apple-system, BlinkMacSystemFont, sans-serif",
    size: 11,
    weight: 400,
    lineHeight: 1.4,
  },
};

// Canvas for text measurement
let measureCanvas: HTMLCanvasElement | OffscreenCanvas | null = null;
let measureCtx: CanvasRenderingContext2D | OffscreenCanvasRenderingContext2D | null = null;

function getContext() {
  if (!measureCtx) {
    if (typeof OffscreenCanvas !== "undefined") {
      measureCanvas = new OffscreenCanvas(1, 1);
      measureCtx = measureCanvas.getContext("2d");
    } else if (typeof document !== "undefined") {
      measureCanvas = document.createElement("canvas");
      measureCtx = measureCanvas.getContext("2d");
    }
  }
  return measureCtx;
}

/**
 * Measure text width and height for a given string and font configuration
 */
export function measureText(
  text: string,
  fontConfig: FontConfig
): { width: number; height: number } {
  if (!text) return { width: 0, height: 0 };

  const cacheKey = `${text}|${fontConfig.size}|${fontConfig.weight}`;
  const cached = measureCache.get(cacheKey);
  if (cached) return cached;

  const ctx = getContext();
  if (!ctx) {
    // Fallback: estimate based on character count
    const width = text.length * fontConfig.size * 0.6;
    const height = fontConfig.size * fontConfig.lineHeight;
    return { width, height };
  }

  ctx.font = `${fontConfig.weight} ${fontConfig.size}px ${fontConfig.family}`;
  const metrics = ctx.measureText(text);

  const width = metrics.width;
  const height = fontConfig.size * fontConfig.lineHeight;

  const result = { width, height };
  measureCache.set(cacheKey, result);
  return result;
}

/**
 * Measure multi-line text with word wrapping
 */
export function measureMultilineText(
  text: string,
  maxWidth: number,
  fontConfig: FontConfig
): { width: number; height: number; lines: string[] } {
  if (!text) return { width: 0, height: 0, lines: [] };

  const words = text.split(" ");
  const lines: string[] = [];
  let currentLine = "";
  let maxLineWidth = 0;

  const ctx = getContext();
  if (ctx) {
    ctx.font = `${fontConfig.weight} ${fontConfig.size}px ${fontConfig.family}`;
  }

  for (const word of words) {
    const testLine = currentLine ? `${currentLine} ${word}` : word;
    const testWidth = ctx
      ? ctx.measureText(testLine).width
      : testLine.length * fontConfig.size * 0.5;

    if (testWidth > maxWidth && currentLine) {
      lines.push(currentLine);
      const lineWidth = ctx
        ? ctx.measureText(currentLine).width
        : currentLine.length * fontConfig.size * 0.5;
      maxLineWidth = Math.max(maxLineWidth, lineWidth);
      currentLine = word;
    } else {
      currentLine = testLine;
    }
  }

  if (currentLine) {
    lines.push(currentLine);
    const lineWidth = ctx
      ? ctx.measureText(currentLine).width
      : currentLine.length * fontConfig.size * 0.5;
    maxLineWidth = Math.max(maxLineWidth, lineWidth);
  }

  const height = lines.length * fontConfig.size * fontConfig.lineHeight;

  return { width: maxLineWidth, height, lines };
}

/**
 * Calculate optimal node size based on content
 */
export function calculateNodeSize(
  title: string,
  technology?: string,
  description?: string,
  nodeType: string = "container"
): { width: number; height: number } {
  // Padding based on node type
  const padding = {
    horizontal: nodeType === "person" ? 24 : 20,
    vertical: nodeType === "person" ? 20 : 16,
  };

  // Measure title
  const titleMeasure = measureText(title, FONT_CONFIG.title);

  // Measure technology tag
  const techMeasure = technology
    ? measureText(`[${technology}]`, FONT_CONFIG.technology)
    : { width: 0, height: 0 };

  // Calculate max width for description wrapping
  const contentWidth = Math.max(titleMeasure.width, techMeasure.width);
  const maxDescWidth = Math.max(contentWidth, 300); // Increased min width for wrapping to 300px

  // Measure description with wrapping (max 2 lines)
  const descMeasure = description
    ? measureMultilineText(description, maxDescWidth, FONT_CONFIG.description)
    : { width: 0, height: 0, lines: [] };

  // Limit description to reasonable max (e.g. 10 lines) instead of 2
  // We want to show as much as possible without breaking the layout completely
  const descHeight = descMeasure.height; // Use full measured height

  // Calculate total dimensions
  const totalContentWidth = Math.max(titleMeasure.width, techMeasure.width, descMeasure.width);

  const width = totalContentWidth + padding.horizontal * 2;

  const height =
    titleMeasure.height +
    (techMeasure.height > 0 ? techMeasure.height + 4 : 0) +
    (descHeight > 0 ? descHeight + 8 : 0) +
    padding.vertical * 2 +
    24; // Icon space

  // Apply min/max constraints based on node type
  // Significantly relaxed constraints to fit content
  const constraints = {
    person: { minWidth: 140, maxWidth: 300, minHeight: 180, maxHeight: 400 },
    system: { minWidth: 150, maxWidth: 400, minHeight: 100, maxHeight: 400 },
    container: { minWidth: 160, maxWidth: 400, minHeight: 100, maxHeight: 400 },
    component: { minWidth: 140, maxWidth: 400, minHeight: 90, maxHeight: 400 },
    datastore: { minWidth: 140, maxWidth: 300, minHeight: 100, maxHeight: 400 },
    queue: { minWidth: 140, maxWidth: 300, minHeight: 100, maxHeight: 400 },
  };

  const nodeConstraints =
    constraints[nodeType as keyof typeof constraints] || constraints.container;

  return {
    width: Math.max(nodeConstraints.minWidth, Math.min(nodeConstraints.maxWidth, Math.ceil(width))),
    height: Math.max(
      nodeConstraints.minHeight,
      Math.min(nodeConstraints.maxHeight, Math.ceil(height))
    ),
  };
}

/**
 * Clear measurement cache (call when font styles change)
 */
export function clearMeasureCache(): void {
  measureCache.clear();
}

File: pkg/export/dot/text_measure.go
// Package dot provides text measurement utilities for accurate node sizing.
//
// This implements FAANG-level text measurement using font metrics.

package dot

import (
	"math"
	"strings"
)

// FontMetrics holds font measurement configuration.
type FontMetrics struct {
	FontSize   float64 // Font size in points
	FontWeight string  // "normal", "bold"
	FontFamily string  // "Arial", etc.
	LineHeight float64 // Line height multiplier (typically 1.2)
}

// DefaultFontMetrics returns default font metrics for node labels.
func DefaultFontMetrics() FontMetrics {
	return FontMetrics{
		FontSize:   12.0,
		FontWeight: "normal",
		FontFamily: "Arial",
		LineHeight: 1.2,
	}
}

// TitleFontMetrics returns font metrics for node titles (bold, larger).
func TitleFontMetrics() FontMetrics {
	return FontMetrics{
		FontSize:   14.0,
		FontWeight: "bold",
		FontFamily: "Arial",
		LineHeight: 1.2,
	}
}

// MeasureText measures text width using font metrics.
// This is a simplified implementation using character-based estimation.
// For production, consider using a proper font rendering library.
func MeasureText(text string, metrics FontMetrics) float64 {
	if text == "" {
		return 0
	}

	// Character width estimation based on Arial metrics
	// Arial 12pt: average char width ~7.2px, bold ~8.0px
	// Arial 14pt: average char width ~8.4px, bold ~9.6px
	charWidth := 7.2
	if metrics.FontSize == 14.0 {
		charWidth = 8.4
		if metrics.FontWeight == "bold" {
			charWidth = 9.6
		}
	} else if metrics.FontWeight == "bold" {
		charWidth = 8.0
	}

	// Scale by font size
	charWidth *= metrics.FontSize / 12.0

	// Measure actual text (account for variable character widths)
	width := 0.0
	for _, r := range text {
		// Adjust for character width variations
		switch r {
		case 'i', 'l', 'I', '1', '|':
			width += charWidth * 0.4
		case 'm', 'w', 'M', 'W':
			width += charWidth * 1.3
		case ' ', '\t':
			width += charWidth * 0.5
		default:
			width += charWidth
		}
	}

	return width
}

// WrapText wraps text to fit within maxWidth, returning lines and dimensions.
func WrapText(text string, maxWidth float64, metrics FontMetrics) (lines []string, width, height float64) {
	if text == "" {
		return []string{}, 0, 0
	}

	words := strings.Fields(text)
	if len(words) == 0 {
		return []string{}, 0, 0
	}

	var currentLine strings.Builder
	currentLine.WriteString(words[0])
	maxLineWidth := MeasureText(words[0], metrics)

	for i := 1; i < len(words); i++ {
		word := words[i]
		testLine := currentLine.String() + " " + word
		testWidth := MeasureText(testLine, metrics)

		if testWidth > maxWidth && currentLine.Len() > 0 {
			// Current line is full, start new line
			lines = append(lines, currentLine.String())
			if testWidth > maxLineWidth {
				maxLineWidth = testWidth
			}
			currentLine.Reset()
			currentLine.WriteString(word)
		} else {
			// Add word to current line
			if currentLine.Len() > 0 {
				currentLine.WriteString(" ")
			}
			currentLine.WriteString(word)
			if testWidth > maxLineWidth {
				maxLineWidth = testWidth
			}
		}
	}

	// Add last line
	if currentLine.Len() > 0 {
		lines = append(lines, currentLine.String())
	}

	// Calculate dimensions
	width = maxLineWidth
	lineHeight := metrics.FontSize * metrics.LineHeight
	height = float64(len(lines)) * lineHeight

	return lines, width, height
}

// MeasureNodeContent measures all content in a node (title, technology, description).
func MeasureNodeContent(elem *Element) (width, height float64) {
	// Measure title (bold, 14pt)
	titleMetrics := TitleFontMetrics()
	titleWidth := MeasureText(elem.Title, titleMetrics)
	titleHeight := titleMetrics.FontSize * titleMetrics.LineHeight

	// Measure technology tag (normal, 12pt)
	var techWidth float64
	var techHeight float64
	if elem.Technology != "" {
		techText := "[" + elem.Technology + "]"
		techMetrics := DefaultFontMetrics()
		techWidth = MeasureText(techText, techMetrics)
		techHeight = techMetrics.FontSize * techMetrics.LineHeight
	}

	// Measure description (wrapped, 12pt)
	var descWidth, descHeight float64
	if elem.Description != "" {
		// Wrap description to reasonable width (max 300px)
		descMetrics := DefaultFontMetrics()
		_, descWidth, descHeight = WrapText(elem.Description, 300, descMetrics)
	}

	// Calculate total width (max of title, tech, desc)
	width = math.Max(titleWidth, math.Max(techWidth, descWidth))

	// Add padding (20px horizontal, 15px vertical)
	width += 40                                         // 20px each side
	height = titleHeight + techHeight + descHeight + 30 // 15px top + 15px bottom

	// Apply min/max constraints
	minWidth, maxWidth := 180.0, 500.0
	minHeight, maxHeight := 100.0, 300.0

	// Adjust based on kind
	switch elem.Kind {
	case "person":
		minWidth, minHeight = 200.0, 180.0
	case "system":
		minWidth, minHeight = 220.0, 140.0
	case "container":
		minWidth, minHeight = 200.0, 120.0
	case "component":
		minWidth, minHeight = 180.0, 100.0
	case "datastore", "queue":
		minWidth, minHeight = 200.0, 100.0
	}

	if width < minWidth {
		width = minWidth
	}
	if width > maxWidth {
		width = maxWidth
	}
	if height < minHeight {
		height = minHeight
	}
	if height > maxHeight {
		height = maxHeight
	}

	return width, height
}
